flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/ExecutionContext.java

```diff
@@ -391,55 +431,119 @@ private static Executor lookupExecutor(
+		if (noInheritedState) {
+			//--------------------------------------------------------------------------------------------------------------
+			// Step.2 Create modules and load them into the TableEnvironment.
+			//--------------------------------------------------------------------------------------------------------------
+			// No need to register the modules info if already inherit from the same session.
+			Map<String, Module> modules = new LinkedHashMap<>();
+			environment.getModules().forEach((name, entry) ->
+					modules.put(name, createModule(entry.asMap(), classLoader))
+			);
+			if (!modules.isEmpty()) {
+				// unload core module first to respect whatever users configure
+				tableEnv.unloadModule(CoreModuleDescriptorValidator.MODULE_TYPE_CORE);
+				modules.forEach(tableEnv::loadModule);
+			}
+
+			//--------------------------------------------------------------------------------------------------------------
+			// Step.3 create user-defined functions and temporal tables then register them.
+			//--------------------------------------------------------------------------------------------------------------
+			// No need to register the functions if already inherit from the same session.
+			registerFunctions();
+
+			//--------------------------------------------------------------------------------------------------------------
+			// Step.4 Create catalogs and register them.
+			//--------------------------------------------------------------------------------------------------------------
+			// No need to register the catalogs if already inherit from the same session.
+			initializeCatalogs();
+		}
+	}
+
+	private void createTableEnvironment(SessionState sessionState) {
+		final EnvironmentSettings settings = environment.getExecution().getEnvironmentSettings();
+		final boolean noInheritedState = sessionState == null;
+		CatalogManager catalogManager;
+		ModuleManager moduleManager;
+		FunctionCatalog functionCatalog;
+		if (noInheritedState) {
+			// Step 1.1 Initialize the CatalogManager if required.
+			catalogManager = new CatalogManager(
+					settings.getBuiltInCatalogName(),
+					new GenericInMemoryCatalog(
+							settings.getBuiltInCatalogName(),
+							settings.getBuiltInDatabaseName()));
+			// Step 1.2 Initialize the ModuleManager if required.
+			moduleManager = new ModuleManager();
+			// Step 1.3 Initialize the FunctionCatalog if required.
+			functionCatalog = new FunctionCatalog(catalogManager, moduleManager);
+			// Step 1.4 Set up session state.
+			this.sessionState = SessionState.of(catalogManager, moduleManager, functionCatalog);
+		} else {
+			catalogManager = sessionState.catalogManager;
+			moduleManager = sessionState.moduleManager;
+			functionCatalog = sessionState.functionCatalog;
+			// Set up session state.
+			this.sessionState = sessionState;
+		}
+		createTableEnvironment(settings, catalogManager, moduleManager, functionCatalog);
+	}
+
+	private void createTableEnvironment(
```

Comment:
you can put merge this method with previous one, then you don't have to assign 3 parameters in the previous method and pass them to this one.